flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/collect/CollectSinkFunction.java

```diff
@@ -0,0 +1,425 @@
+			getRuntimeContext().addAccumulator(finalResultOffsetAccumulatorName, offsetAccumulator);
+		} finally {
+			bufferedResultsLock.unlock();
+		}
+	}
+
+	@Override
+	public void notifyCheckpointComplete(long checkpointId) {
+		lastCheckpointedOffset = uncompletedCheckpointMap.get(checkpointId);
+		uncompletedCheckpointMap.headMap(checkpointId + 1).clear();
+	}
+
+	public void setOperatorEventGateway(OperatorEventGateway eventGateway) {
+		this.eventGateway = eventGateway;
+	}
+
+	/**
+	 * The thread that runs the socket server.
+	 */
+	private class ServerThread extends Thread {
+
+		private final ServerSocket serverSocket;
+
+		private boolean running;
+
+		private Socket connection;
+		private DataInputViewStreamWrapper inStream;
+		private DataOutputViewStreamWrapper outStream;
+
+		private ServerThread() throws Exception {
+			this.serverSocket = new ServerSocket(0, 0, getBindAddress());
+			this.running = true;
+		}
+
+		@Override
+		public void run() {
+			while (running) {
+				try {
+					if (connection == null) {
+						// waiting for coordinator to connect
+						connection = serverSocket.accept();
+						inStream = new DataInputViewStreamWrapper(this.connection.getInputStream());
+						outStream = new DataOutputViewStreamWrapper(this.connection.getOutputStream());
+						LOG.info("Coordinator connection received");
+					}
+
+					CollectCoordinationRequest request = new CollectCoordinationRequest(inStream);
+					String requestVersion = request.getVersion();
+					// client acknowledges that it has successfully received results before this offset,
+					// we can safely throw away results before this offset
+					long requestOffset = request.getOffset();
+					if (LOG.isDebugEnabled()) {
+						LOG.debug(
+							"Request received, version = " + requestVersion + ", offset = " + requestOffset);
+						LOG.debug(
+							"Expecting version = " + version + ", offset = " + offset);
+					}
+
+					if (!version.equals(requestVersion) || requestOffset < offset) {
+						// invalid request
+						LOG.warn("Invalid request. Received version = " + requestVersion +
+							", offset = " + requestOffset + ", while expected version = "
+							+ version + ", offset = " + offset);
+						sendBackResults(Collections.emptyList());
+						continue;
+					}
+
+					// valid request, sending out results
+					List<IN> results;
+					bufferedResultsLock.lock();
+					try {
+						int oldSize = bufferedResults.size();
+						int ackedNum = Math.min((int) (requestOffset - offset), oldSize);
+						int nextBatchSize = Math.min(ackedNum + maxResultsPerBatch, oldSize) - ackedNum;
+
+						if (LOG.isDebugEnabled()) {
+							LOG.debug("Preparing " + nextBatchSize + " results");
+						}
+
+						// drop acked results
+						for (int i = 0; i < ackedNum; i++) {
+							bufferedResults.removeFirst();
+							offset++;
+						}
+
+						// prepare next result batch
+						results = new ArrayList<>(bufferedResults.subList(0, nextBatchSize));
```

Comment:
`results = bufferedResults.subList(0, nextBatchSize);` ?

Reply:
No. `subList` just returns a "slice" of the original list. This might cause concurrent modifying.